File: dbxml/examples/java/gettingStarted/queryWithContext.java
//
// See the file LICENSE for redistribution information.
//
// Copyright (c) 2004,2009 Oracle.  All rights reserved.
//
//

package dbxml.gettingStarted;

import java.io.*;
import com.sleepycat.dbxml.*;

class queryWithContext
{

    private static String theContainer = "namespaceExampleData.dbxml";
    private static String collection = "collection('namespaceExampleData.dbxml')";
    private static void usage() {
	String usageMessage = "\nThis program illustrates how to query for documents that require namespace\n";
	usageMessage += "usage in the query. You should run exampleLoadContainer before running this example.\n";
	usageMessage += "You are only required to pass this command the path location of the database\n";
	usageMessage += "environment that you specified when you loaded the examples data:\n\n";

	usageMessage += "\t-h <dbenv directory>\n";

	usageMessage += "For example:\n";
	usageMessage += "\tjava com.sleepycat.dbxml.examples.gettingStarted.queryWithContext -h examplesEnvironment\n";

	System.out.println(usageMessage);
	System.exit( -1 );
    }

    //Performs a query against a document using an XmlQueryContext.
    private static void doContextQuery( XmlManager mgr, String query,
					XmlQueryContext context )
	throws Throwable {
	String fullQuery = collection + query;
	//Perform a single query against the referenced container using
	// the referenced context.
	System.out.println("Exercising query: '" + fullQuery + "'.");
	System.out.println("Return to continue: ");
	System.in.read();

	//Perform the query
	XmlResults results = mgr.query(fullQuery, context, null);
	//Iterate over the results with an XmlValue
	XmlValue value = results.next();
	while (value != null) {
            System.out.println(value.asString());
            value = results.next();
	}
	System.out.println(results.size() + " results returned for query '"
			   + fullQuery + "'.");
	results.delete();
    }

    //Utility function to clean up objects, exceptions or not,
    // containers and environments must be closed.
    private static void cleanup(myDbEnv env, XmlContainer openedContainer) {
	try {
	    if (openedContainer != null)
		openedContainer.delete();
	    if (env != null)
		env.cleanup();
	} catch (Exception e) {
	    // ignore exceptions on close
	}
    }

    public static void main(String args[])
	throws Throwable {

	File path2DbEnv = null;

	for(int i = 0; i < args.length; ++i) {
            if (args[i].startsWith("-")) {
		switch(args[i].charAt(1)) {
		case 'h':
		    path2DbEnv = new File( args[++i] );
		    break;
		default:
                    usage();
		}
            }
	}

	if (path2DbEnv == null || ! path2DbEnv.isDirectory()) {
            usage();
	}

	myDbEnv env = null;
	XmlContainer openedContainer = null;
	try {
	    //Create an environment
	    env = new myDbEnv(path2DbEnv);
	    XmlManager theMgr = env.getManager();

	    //Open a container in that environment
	    openedContainer =
		theMgr.openContainer(theContainer);

            //Create a context and use it to declare the namespaces
            XmlQueryContext context = theMgr.createQueryContext();

            context.setNamespace("fruits", "http://groceryItem.dbxml/fruits");
            context.setNamespace("vegetables",
				 "http://groceryItem.dbxml/vegetables");
            context.setNamespace("desserts", "http://groceryItem.dbxml/desserts");

            //Set a variable
            context.setVariableValue("aDessert",
				     new XmlValue("Blueberry Boy Bait"));

            //Perform the queries

            //Find all the Vendor documents in the database. Vendor documents do
            // not use namespaces, so this query returns documents.
            doContextQuery(theMgr, "/vendor", context);

            //Find the product document for "Lemon Grass".
	    // This query returns no documents
            // because a namespace prefix is not identified for the 'item' node.
            doContextQuery(theMgr,
			   "/item/product[.=\"Lemon Grass\"]", context);

            //Find the product document for "Lemon Grass" using the namespace
	    // prefix 'fruits'. This query successfully returns a document.
            doContextQuery(theMgr,
			   "/fruits:item/product[.=\"Lemon Grass\"]", context);

            //Find all the vegetables
            doContextQuery(theMgr, "/vegetables:item", context);

            //Find the  dessert called Blueberry Boy Bait.
            // Note the use of a variable.
            doContextQuery(theMgr,
			   "/desserts:item/product[.=$aDessert]", context);

       } catch (Exception e) {
           System.err.println("Error performing query against " + theContainer );
           System.err.println("   Message: " + e.getMessage() );
           throw e;
       }
	finally {
	    cleanup(env, openedContainer);
	}
    } //End main
}

File: dbxml/examples/perl/gettingStarted/queryWithContext.pl


use strict;
use warnings;
use Sleepycat::DbXml 'simple';


sub usage()
{
    warn  <<EOM;
This program illustrates how to query for documents that require namespace
usage in the query.  You should pre-load the container using
loadExamplesData.[sh|cmd] before running this example.  You are only required
to pass this command the path location of the database environment that you
specified when you pre-loaded the examples data:

    -h <dbenv directory>
EOM
        exit( -1 );
}

sub doContextQuery($$$$)
{
    my $mgr = shift ;
	my $cname = shift ;
    my $query = shift ;
    my $context = shift ;

    # //////////////////////////////////////////////////////////////////////
    # ////  Performs a simple query (with context) against the ///////
    # ////  provided container.                                      ///////
    # //////////////////////////////////////////////////////////////////////
	
    # /// some defensive code eliminated for clarity //
	
    #  Perform the query. Result type is by default Result Document
    my $fullQuery = "collection('$cname')$query";
    eval {
	print "Exercising query '$fullQuery'\n";
	print "Return to continue: ";
	<>;
	
	my $results = $mgr->query($fullQuery, $context);
	my $value;
	while( $results->next($value) )
	{
            #  Obtain the value as a string and print it to the console
	    print "$value\n";
	}
	
	print $results->size() . " objects returned for expression '$fullQuery'\n" ;
	
    };
    if (my $e = catch std::exception)
    {
        warn "Query $fullQuery failed\n";
        warn $e->what() . "\n";
        exit( -1 );
    }
    elsif ($@)
    {
        warn "Query $fullQuery failed\n";
        warn $@;
        exit( -1 );
    }
    
}

my $path2DbEnv;
my $theContainer = "namespaceExampleData.dbxml";
while (@ARGV && $ARGV[0] =~ /^-/ )
{
    my $opt = shift @ARGV ;

    if ($opt =~ s/-h//)
      { $path2DbEnv = $opt || shift }
    else
      { usage() }
}

usage()
  if ! $path2DbEnv ;

eval {
    # open a container in the db environment
    my $env = new DbEnv(0);
    $env->set_cachesize(0, 64 * 1024 * 1024, 1);
    print "opening $path2DbEnv\n";
    $env->open($path2DbEnv,
                Db::DB_INIT_MPOOL|Db::DB_CREATE|Db::DB_INIT_LOCK|Db::DB_INIT_LOG|Db::DB_INIT_TXN);
    my $theMgr = new XmlManager($env);

    my $containerTxn = $theMgr->createTransaction();
    my $container = $theMgr->openContainer($containerTxn, $theContainer);
    $containerTxn->commit();


  # Create a context and declare the namespaces
  my $context = $theMgr->createQueryContext();
  $context->setNamespace( "fruits" => "http://groceryItem.dbxml/fruits");
  $context->setNamespace( "vegetables" => "http://groceryItem.dbxml/vegetables");
  $context->setNamespace( "desserts" => "http://groceryItem.dbxml/desserts");

  # Set a variable
  $context->setVariableValue("aDessert", "Blueberry Boy Bait");

  # Perform the queries

  # Find all the Vendor documents in the database. Vendor documents do
  #  not use namespaces, so this query returns documents.
  doContextQuery( $theMgr, $container->getName(), "/vendor", $context );

  # Find the product document for "Lemon Grass". This query returns no documents
  #  because a namespace prefix is not identified for the 'item' node.
  doContextQuery( $theMgr, $container->getName(),
		  '/item/product[text()="Lemon Grass"]', $context);

  # Find the product document for "Lemon Grass" using the namespace prefix 'fruits'. This
  #  query successfully returns a document.
  doContextQuery( $theMgr, $container->getName(),
		  '/fruits:item/product[text()="Lemon Grass"]', $context);

  # Find all the vegetables
  doContextQuery( $theMgr, $container->getName(), "/vegetables:item", $context);

  # Find the  dessert called Blueberry Boy Bait
  #  Note the use of a variable
  doContextQuery( $theMgr, $container->getName(),
		  '/desserts:item/product[text()=$aDessert]', $context);

};
if (my $e = catch std::exception)
{
    warn "Query failed\n";
    warn $e->what() . "\n";
    exit( -1 );
}
elsif ($@)
{
    warn "Query failed\n";
    warn $@;
    exit( -1 );
}

exit 0 ;

